packages/plugins/typescript/urql-graphcache/src/index.ts

```diff
@@ -0,0 +1,244 @@
+import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
+import { GraphQLSchema, isWrappingType, Kind, GraphQLWrappingType, TypeNode, GraphQLAbstractType } from 'graphql';
+import { UrqlGraphCacheConfig } from './config';
+import { baseTypes, imports } from './constants';
+
+type GraphQLFlatType = Exclude<TypeNode, GraphQLWrappingType>;
+
+const unwrapType = (type: null | undefined | TypeNode): GraphQLFlatType | null => {
+  if (isWrappingType(type)) {
+    return unwrapType(type.ofType);
+  }
+
+  return type || null;
+};
+
+function constructType(
+  typeNode: TypeNode,
+  schema: GraphQLSchema,
+  nullable = true,
+  allowString = false
+): string {
+  switch (typeNode.kind) {
+    case 'ListType': {
+      return nullable
+        ? `Maybe<Array<${constructType(typeNode.type, schema, false, allowString)}>>`
+        : `Array<${constructType(typeNode.type, schema, false, allowString)}>`;
+    }
+    case 'NamedType': {
+      const type = schema.getType(typeNode.name.value);
+      if (!type.astNode) {
+        return nullable ? `Maybe<Scalars['${typeNode.name.value}']}>` : `Scalars['${typeNode.name.value}']`;
+      }
+
+      switch (type.astNode.kind) {
+        case 'EnumTypeDefinition':
+        case 'UnionTypeDefinition':
+        case 'InputObjectTypeDefinition':
+        case 'ObjectTypeDefinition':
+        case 'ScalarTypeDefinition': {
+          const finalType = `RequireFields<${typeNode.name.value}, '__typename'>${allowString ? ' | string' : ''}`;
```

Comment:
Opted for using `RequireFields` here so people that don't force __typename to required in the `typescript` plugin don't get errors.

Reply:
Cool. This is an alternative for requiring users to set `nonOptionalTypename: true` explicitly?